### PR TITLE
 OS-8334 Bring OpenSSL 3 to the platform

### PR DESCRIPTION
**Pull request by danmcd**
Many, many, many changes here, as described in the OS ticket.

### CODE REVIEW

File: openssl3/install-sfw [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+mkdir -p ${DESTDIR}/etc/openssl/private
+mkdir -p ${DESTDIR}/etc/sfw
+chmod 0700 ${DESTDIR}/etc/openssl/private
+_install N apps/openssl.cnf ${DESTDIR}/etc/openssl/openssl.cnf 0644
+_install L ../openssl ${DESTDIR}/etc/sfw/openssl
+
+# Install header files
+mkdir -p ${INCDIR}/openssl
+cd include/openssl
+# XXX KEBE SAYS THIS MAY NEED MORE FIXES...
```
Comment (danmcd) [inline openssl3/install-sfw:88]: As of now, I can probably strike this comment.

**Review by danmcd — COMMENTED** [review]
There are other KEBE comments in here as well.  I am aware...

File: openssl3/install-sfw [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    ${USRLIBDIR}/libcrypto-smartos.so.${LIBVER}
+_install L ../../lib/libcrypto-smartos.so ${USRLIBDIR}/libcrypto-smartos.so
+_install L ../../lib/libcrypto-smartos.so ${USRLIBDIR}/libsunw_crypto.so
+_install L ../../lib/libssl-smartos.so.${LIBVER} \
+    ${USRLIBDIR}/libssl-smartos.so.${LIBVER}
+_install L ../../lib/libssl-smartos.so ${USRLIBDIR}/libssl-smartos.so
+_install L ../../lib/libssl-smartos.so ${USRLIBDIR}/libsunw_ssl.so
+
+# Install commands
+## XXX KEBE ASKS, put 64-bit version in plain bin/openssl ? (for now no)
```
Comment (bahamat) [inline openssl3/install-sfw:70]: This comment should also be removed, or turned into a more generalized explanation.

File: openssl3/Makefile [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+SHELL =		/bin/bash
+BUILD64 =	yes
+UNPACK_SENTINEL =	Configure
+AUTOCONF_OUT =	crypto/opensslconf.h
+
+MT_CPPFLAGS =	-D_REENTRANT
+SHARED_CFLAGS =	-fPIC -DPIC
+SHARED_LDFLAGS =	-fPIC -shared
+
+# XXX KEBE SAYS consider losing the PKCS#11 engine, given 3.0 deprecates it in
```
Comment (bahamat) [inline openssl3/Makefile:39]: And here?

File: openssl3/install-sfw-64 [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    ${USRLIBDIR}/libsunw_crypto.so
+_install L ../../../lib/amd64/libssl-smartos.so.${LIBVER} \
+    ${USRLIBDIR}/libssl-smartos.so.${LIBVER}
+_install L ../../../lib/amd64/libssl-smartos.so ${USRLIBDIR}/libssl-smartos.so
+_install L ../../../lib/amd64/libssl-smartos.so ${USRLIBDIR}/libsunw_ssl.so
+
+# Install commands
+_install E apps/openssl ${BINDIR}/openssl 0555
+_install L ../../../bin/amd64/openssl ${SFWBINDIR}/openssl
+## XXX KEBE ASKS, put 64-bit version in plain bin/openssl ? (for now no)
```
Comment (bahamat) [inline openssl3/install-sfw-64:75]: Last one here.

File: openssh/Makefile [hunk 5]
```diff
@@ -22,7 +22,7 @@
 # Copyright 2019 Joyent, Inc.
```
Comment (bahamat) [inline openssh/Makefile:22]: Copyright update

**Review by bahamat — CHANGES_REQUESTED** [review]
Just a few minor nits here.

File: openssl3/install-sfw (hunk 2 above)
Comment (danmcd) [reply openssl3/install-sfw:70]: There's an actual open question here that I haven't totally decided yet.  Today we deliver 32-bit openssl(1) in /usr/bin/openssl, and 64-bit openssl(1) in /usr/bin/amd64/openssl. I'm not sure if there's benefit (like there is with ls(1)) to defaulting to 32-bit, or whether or not we should just go 64-bit-only.

File: openssl3/Makefile (hunk 3 above)
Comment (danmcd) [reply openssl3/Makefile:39]: Here I've made a decision, so I need to clarify.  Will fix.

File: openssl3/install-sfw-64 (hunk 4 above)
Comment (danmcd) [reply openssl3/install-sfw-64:75]: See above...

File: openssh/Makefile (hunk 5 above)
Comment (danmcd) [reply openssh/Makefile:22]: Will fix.

File: openssh/Makefile [hunk 6]
```diff
@@ -22,7 +22,7 @@
...
 #
 
-VER =	openssh-portable-8.6p1
+VER =	openssh-8.8p1
```
Comment (bahamat) [inline openssh/Makefile:25]: Should this still be openssh-portable?

File: openssh/Patches/0002-PAM-Support.patch [hunk 7]
```diff
@@ -31,16 +31,15 @@ index 4d1910fe..ac105537 100644
  
  	/* Standard Options */
  	if (options->num_host_key_files == 0) {
-@@ -1309,8 +1314,17 @@ process_server_config_line_depth(ServerOptions *options, char *line,
+@@ -1329,8 +1334,16 @@ process_server_config_line_depth(ServerOptions *options, char *line,
  	switch (opcode) {
  	/* Portable-specific options */
  	case sUsePAM:
 +#ifdef SET_USE_PAM
 +		/* UsePAM is always on and not configurable on Solaris */
```
Comment (bahamat) [inline openssh/Patches/0002-PAM-Support.patch:39]: While you're here..change this to illumos?

File: openssh/Makefile (hunk 6 above)
Comment (danmcd) [reply openssh/Makefile:25]: The download from openssh.org doesn't say portable, but I can rename and fix this, sure.

File: openssh/Patches/0002-PAM-Support.patch (hunk 7 above)
Comment (danmcd) [reply openssh/Patches/0002-PAM-Support.patch:39]: Sure why not?

Comment (danmcd) [reply openssh/Patches/0002-PAM-Support.patch:39]: Oh, ummm, this patch actually from Oracle, so I'd prefer not to mess with it.

Comment (bahamat) [reply openssh/Patches/0002-PAM-Support.patch:39]: Ok, I'm fine with that.

File: openssl3/install-sfw (hunk 2 above)
Comment (bahamat) [reply openssl3/install-sfw]: I'd say the cautious route is to go 32-bit now, and we can always change it later. Eventually the whole platform will become 64-bit, but that's a problem for a different day.

Comment (danmcd) [reply openssl3/install-sfw]: I agree, and the "code review feedback" push e430039a1f confirms the "for now no" and makes it a real decision (that lessens surprise from OpenSSL 1.0.2).  Thank you!

File: node.js/Makefile [hunk 8]
```diff
@@ -29,6 +29,13 @@ include ../Makefile.defs
 NODE_ROOT =	$(DESTDIR)/usr/node/$(MAJOR_VER)
```
Comment (papertigers) [inline node.js/Makefile:29]: Update the copyright above

File: openssl3/install-sfw [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+LIBDIR=${DESTDIR}/lib
+BINDIR=${DESTDIR}/usr/bin
+USRLIBDIR=${DESTDIR}/usr/lib
+INCDIR=${DESTDIR}/usr/include
+
+mkdir -p $LIBDIR
+mkdir -p $BINDIR
+mkdir -p $USRLIBDIR
+mkdir -p $INCDIR
```
Comment (papertigers) [inline openssl3/install-sfw:40]: Technically these should be quoted I believe.  I'm not a bash expert but did you run this through shell check tool?

File: openssl3/install-sfw [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+_install L ../../lib/libssl-smartos.so ${USRLIBDIR}/libsunw_ssl.so
+
+# Install commands
+_install P apps/CA.pl ${BINDIR}/CA.pl 0555
+_install E apps/openssl ${BINDIR}/openssl 0555
+mkdir -p ${DESTDIR}/usr/sfw/bin
+_install L ../../bin/openssl ${DESTDIR}/usr/sfw/bin/openssl
+
+# Install etc
+mkdir -p ${DESTDIR}/etc/openssl
```
Comment (papertigers) [inline openssl3/install-sfw:76]: Same with these

Comment (danmcd) [reply openssl3/install-sfw:76]: Same answer as above.

**Review by danmcd — COMMENTED** [review]
Will fix the copyright, but won't shell-clean the OpenSSL 3 forward-ports of the installer scripts.  I think we should wait until OpenSSL 1x is gone before that.

**Review by bahamat — APPROVED** [review]

**Review by papertigers — APPROVED** [review]

**Pull request merged**